consensus/src/sync/light/sync.rs

```diff
@@ -0,0 +1,177 @@
+use std::collections::{HashMap, VecDeque};
+
+use std::sync::Arc;
+use std::task::Waker;
+
+use futures::{future::BoxFuture, stream::FuturesUnordered, FutureExt};
+use nimiq_block::Block;
+use nimiq_light_blockchain::LightBlockchain;
+use nimiq_network_interface::peer::CloseReason;
+use nimiq_network_interface::request::RequestError;
+use nimiq_zkp_component::types::{Error, ZKPRequestEvent};
+use nimiq_zkp_component::zkp_component::ZKPComponentProxy;
+use parking_lot::RwLock;
+
+use crate::{messages::Checkpoint, sync::syncer::MacroSync};
+use nimiq_hash::Blake2bHash;
+use nimiq_network_interface::network::{Network, SubscribeEvents};
+
+#[derive(Clone)]
+/// This struct is used to request Epochs IDs (hashes) from other peers
+/// in order to determine their macro chain state relative to us
+pub(crate) struct EpochIds<T> {
+    /// Indicates if the latest epoch id that was queried was found in the peer's chain
+    pub locator_found: bool,
+    /// The most recent epoch ids (hashes)
+    pub ids: Vec<Blake2bHash>,
+    /// The most recent checkpoint block in the latest epoch (if any)
+    pub checkpoint: Option<Checkpoint>,
+    /// The first epoch block number
+    pub first_epoch_number: usize,
+    /// The sender that created this struct
+    pub sender: T,
+}
+
+impl<T> EpochIds<T> {
+    #[inline]
+    pub(crate) fn checkpoint_epoch_number(&self) -> usize {
+        self.first_epoch_number + self.ids.len()
+    }
+
+    #[inline]
+    pub(crate) fn last_epoch_number(&self) -> usize {
+        self.checkpoint_epoch_number().saturating_sub(1)
+    }
+}
+
+/// This struct is used to track all the macro requests sent to a particular peer
+pub struct PeerMacroRequests {
+    /// Number of requests that have been fullfiled
+    completed_requests: usize,
+    /// A Queue used to track the requests that have been sent, and their respective result
+    queued_requests: VecDeque<(Blake2bHash, Option<Block>)>,
+}
+impl PeerMacroRequests {
+    pub fn new() -> Self {
+        Self {
+            completed_requests: 0,
+            queued_requests: VecDeque::new(),
+        }
+    }
+
+    // Pushes a new request into the queue
+    pub fn push_request(&mut self, block_hash: Blake2bHash) {
+        self.queued_requests.push_back((block_hash, None))
+    }
+
+    // Pops a request from the queue
+    pub fn pop_request(&mut self) -> Option<(Blake2bHash, Option<Block>)> {
+        self.queued_requests.pop_back()
+    }
+
+    // Returns true if the request was updated, false in case the request was not found
+    pub fn update_request(&mut self, block: Block) -> bool {
+        let position = self
+            .queued_requests
+            .iter()
+            .position(|(hash, _)| *hash == block.hash());
+
+        if let Some(position) = position {
+            self.queued_requests[position] = (block.hash(), Some(block));
```

Comment:
why?
If it is not none, it means the peer replied two times to the same request, and if that's the case, we just use the last one, overwriting any previous one.
So I don't think this check is necessary, unless we want to really fail if it is Some

Reply:
It is because of what we do later: `self.completed_requests += 1;`
So, if someone sends you the same response twice, you may be marking the completeness flag without actually being complete